pkg/version/version.go

```diff
@@ -52,6 +54,32 @@ func (info Info) String() string {
 	return version
 }
 
+var (
+	reCommitSegment   = regexp.MustCompile(`^g[0-9a-f]{6,10}$`)
+	reCommitIncrement = regexp.MustCompile(`^[0-9a-f]+$`)
+)
+
+// LastSemanticVersion attempts to return a semantic version from the GitVersion - which
+// is either <semver>-<increment>-g<commit> or <semver> on release boundaries.
+func (info Info) LastSemanticVersion() string {
+	version := info.GitVersion
+	parts := strings.Split(version, "-")
+	// strip the modifier
+	if len(parts) > 1 && parts[len(parts)-1] == "dirty" {
+		parts = parts[:len(parts)-1]
+	}
+	// strip the Git commit
+	if len(parts) > 1 && reCommitSegment.MatchString(parts[len(parts)-1]) {
+		parts = parts[:len(parts)-1]
+		// strip a version increment, but only if we found the commit
+		if len(parts) > 1 && reCommitIncrement.MatchString(parts[len(parts)-1]) {
+			parts = parts[:len(parts)-1]
+		}
+	}
+
+	return strings.Join(parts, "-")
+}
```

Comment:
We have not used more than two semantic parts so far.  The possible values
passed to GitVersion are:

v1.1.6
v1.1.6-16-g3498340
v1.1.6-rc1
v1.1.6-rc1-16-g3498340

On Mon, Apr 25, 2016 at 3:49 PM, Scott Dodson <notifications@github.com>
wrote:

> In pkg/version/version.go
> <https://github.com/openshift/origin/pull/8608#discussion_r60976326>:
>
> > +	parts := strings.Split(version, "-")
> > +	// strip the modifier
> > +	if len(parts) > 1 && parts[len(parts)-1] == "dirty" {
> > +		parts = parts[:len(parts)-1]
> > +	}
> > +	// strip the Git commit
> > +	if len(parts) > 1 && reCommitSegment.MatchString(parts[len(parts)-1]) {
> > +		parts = parts[:len(parts)-1]
> > +		// strip a version increment, but only if we found the commit
> > +		if len(parts) > 1 && reCommitIncrement.MatchString(parts[len(parts)-1]) {
> > +			parts = parts[:len(parts)-1]
> > +		}
> > +	}
> > +
> > +	return strings.Join(parts, "-")
> > +}
>
> Is it safe to assume that if we were to split on '-' we'd only include at
> most the first two parts in the image tag?
> v1.1.6 -> v1.1.6
> v1.2.0-rc1-17-g642f0af -> v1.2.0-rc1
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8608/files/8b5309663381ce4cf5cdaaeb737d4d8c7d7f4a0d#r60976326>
>